content/post/2022-04-25-PHP-database-result-iteration.md

```diff
@@ -0,0 +1,164 @@
+---
+title: Best practices in PHP when iterating over database results
+date: 2022-04-25
+tags:
+  - PHP
+  - database
+  - history
+  - patterns
+description: "How the default method for iterating over database results
+is stuck in 2005, goes against best practices and what to do about it."
+---
+
+Many PHP database libraries use a `while` loop with assignment to iterate
+over database results. In this article I'll show why this pattern is so
+pervasive, why it opens the door for errors and what
+you can use instead.
+
+<!--more-->
+
+This is a typical piece of PHP code, using [PDO](https://www.php.net/manual/en/book.pdo.php) for accessing the database and iterating over the result:
+
+```php
+$result = $db->query("SELECT id, name from my_table");
+while ($row = $result->fetch()) {
+    // do something with $row
+}
+```
+
+Many coding styles don't recommend the practice of "assignment inside control structure epxresssion". The [MediaWiki Coding Style](https://www.mediawiki.org/wiki/Manual:Coding_conventions/PHP) even has [a rule that forbids assignments in control structures](https://github.com/wikimedia/mediawiki-tools-codesniffer/blob/master/MediaWiki/Sniffs/ControlStructures/AssignmentInControlStructuresSniff.php), which prevents this style of looping. Which could raise the questions:
+
+- Why disallow assignments in control structures?
+- How do I fix the code to make the coding style checker happy?
+- How did this style of looping evolve and why did the evolution stop?
+
+### Why assignments in control structures are dangerous
+The danger comes from the "overloading" of the equal sign in an expression:
+
+- `$a = $b` is an **assignment**, the value of variable `$b` gets assigned to `$a`. The return type of the expression is the type of `$b`.
+- `$a == $b` is a **comparison**, checking if `$a` and `$b` have the same value after applying [type conversion](https://www.php.net/manual/en/language.types.type-juggling.php). The return type of the expression is a boolean.
+- `$a === $b` is a **comparison**, checking if `$a` and `$b` have the same value and type.The return type of the expression is a boolean.
+
+Control structures like `if` and `while` expect a boolean expression. If you're using an assignment (one equal sign) instead of a comparison (two or three equal signs) the PHP interpreter will first do the assignment and then type-juggle the assigned value into a boolean. This will change the way your code works:
```

Comment:
This is not necessarily a problem:
```PHP
while ( ( $row = $result->fetch() ) !== false ) {
}
```

But this example demonstrates more problems:
* This is just really hard to read. You need to know the execution order to be able to read this: First, `fetch()` is executed → result is stored in `$row` → the surrounding `( … )` return what's in `$row` → it's compared with `false`.
* People tend to forget the extra, type-safe comparison. They memorize the simplified `while ( $row = $result->fetch() )`, which does have all the type-juggling issues described above.

It's not only about accidentally confusing `=` for a comparison.